src/ServiceControl/Recoverability/Retrying/RetryProcessor.cs

```diff
@@ -195,6 +198,57 @@ async Task<int> Stage(RetryBatch stagingBatch, IAsyncDocumentSession session)
             return messages.Length;
         }
 
+        async Task TryStageMessage(FailedMessage message, string stagingId, FailedMessageRetry failedMessageRetry)
+        {
+            try
+            {
+                await StageMessage(message, stagingId).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                if (failedMessageRetry.StageAttempts < 5)
+                {
+                    await IncrementAttemptCounter(failedMessageRetry).ConfigureAwait(false);
+                    //We need to rethrow here to stop the current staging process
+                    throw new Exception($"Error while trying to stage failed message {message.UniqueMessageId}", e);
+                }
+
+                Log.Error($"Failed message {message.UniqueMessageId} reached its staging retry limit ({failedMessageRetry.StageAttempts}) and is going to be removed from the batch.", e);
+
+                await store.AsyncDatabaseCommands.DeleteAsync(FailedMessageRetry.MakeDocumentId(message.UniqueMessageId), null)
+                    .ConfigureAwait(false);
+
+                await domainEvents.Raise(new MessageFailedInStaging
+                {
+                    UniqueMessageId = message.UniqueMessageId
+                }).ConfigureAwait(false);
+
+                //We need to rethrow here to force re-loading the batch
+                throw new Exception($"Error while trying to stage failed message {message.UniqueMessageId}", e);
+            }
+        }
+
+        async Task IncrementAttemptCounter(FailedMessageRetry message)
+        {
+            try
+            {
+                await store.AsyncDatabaseCommands.PatchAsync(message.Id,
+                    new[]
+                    {
+                        new PatchRequest
```

Comment:
Existing document would not have this property set but this way of setting/incrementing it should be fine